WalletWasabi/Services/UpdateManager.cs

```diff
@@ -1,329 +1,252 @@
-		IoHelpers.EnsureContainingDirectoryExists(tmpFilePath);
-		using (var file = File.OpenWrite(tmpFilePath))
-		{
-			await stream.CopyToAsync(file, cancellationToken).ConfigureAwait(false);
-
-			// Closing the file to rename.
-			file.Close();
-		}
-		File.Move(tmpFilePath, filePath);
-	}
 
-	private async Task<ReleaseInfo> GetLatestReleaseFromGithubAsync(CancellationToken cancellationToken)
-	{
-		using HttpRequestMessage message = new(HttpMethod.Get, ReleaseURL);
-		message.Headers.UserAgent.Add(new("WalletWasabi", "2.0"));
-		var response = await _githubHttpClient.SendAsync(message, cancellationToken).ConfigureAwait(false);
+		var asset = assetToDownloadResult.Value;
+		var installerFilePath = await DownloadFileAsync(asset.Uri).ConfigureAwait(false);
 
-		JObject jsonResponse = JObject.Parse(await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false));
+		Logger.LogInfo($"Installer downloaded to: {installerFilePath}");
 
-		string softwareVersion = jsonResponse["tag_name"]?.ToString() ?? throw new InvalidDataException($"Endpoint gave back wrong json data or it's changed.\n{jsonResponse}");
+		// Verify installer hash match the expected one
+		var lines = await File.ReadAllLinesAsync(sha256SumsTask.Result, cancellationToken).ConfigureAwait(false);
+		var installerHash = lines.Select(l => l.Split("  ./", StringSplitOptions.TrimEntries | StringSplitOptions.RemoveEmptyEntries))
+			.Select(a => (Hash: a[0], FileName: a[1]))
+			.FirstOrDefault(a => a.FileName == asset.FileName)
+			.Hash ?? throw new InvalidOperationException($"{asset.FileName} was not found.");;
 
-		// Make sure there are no non-numeric characters (besides '.') in the version string.
-		softwareVersion = string.Concat(softwareVersion.Where(c => char.IsDigit(c) || c == '.').ToArray());
+		await VerifyInstallerHashAsync(installerFilePath, installerHash, cancellationToken).ConfigureAwait(false);
+		Logger.LogInfo("Installer verified successfully");
 
-		Version githubVersion = new(softwareVersion);
+		// Notify that there is an installer ready
+		eventBus.Publish(new NewSoftwareVersionInstallerAvailable(installerFilePath));
+		return;
 
-		// Get all asset names and download URLs to find the correct one.
-		List<JToken> assetsInfo = jsonResponse["assets"]?.Children().ToList() ?? throw new InvalidDataException("Missing assets from response.");
-		List<string> assetDownloadLinks = new();
-		foreach (JToken asset in assetsInfo)
-		{
-			assetDownloadLinks.Add(asset["browser_download_url"]?.ToString() ?? throw new InvalidDataException("Missing download url from response."));
-		}
+		Task<string> DownloadFileAsync(Uri uri) => DownloadAsync(httpClientFactory, uri, installDirectory, cancellationToken);
 
-		return new ReleaseInfo(githubVersion, assetDownloadLinks);
+		Result<(string FileName, Uri Uri), string> Find(string ending) =>
+			releaseInfo.Assets.Select(x => (x.Key, x.Value)).FirstOrDefault(x => x.Key.EndsWith(ending)) is {} found
```

Comment:
This doesn't work on MacOs:`.dmg` will also catch `.arm64.dmg`

Reply:
hahaha I new pinging you was the right thing to do. Good catch